scripts/gen/tpcc.sh

```diff
@@ -118,33 +171,154 @@ then
   SET CLUSTER SETTING kv.snapshot_recovery.max_rate = '128 MiB';
   SET CLUSTER SETTING kv.snapshot_rebalance.max_rate = '128 MiB';
   ";
-
-  echo "importing..."
+  echo "importing ..."
   ./cockroach workload fixtures import tpcc --warehouses="$f_warehouses" --active-warehouses="$f_active" "${pgurls[0]}"
   echo "done importing"
 fi
 
-# if f_inc is not specified, then we set it equal to f_warehouses, this means
-# we just need to run one iteration.
-if [[ $f_inc == 0 ]];
+# Comment out since we don't have a scenario using this at present.
+## if f_inc is not specified, then we set it equal to f_warehouses, this means
+## we just need to run one iteration.
+#if [[ $f_inc == 0 ]];
+#then
+#  f_inc=$f_warehouses
+#fi
+
+# We first run an exploration test round with binary search to find a right
+# range of test boundary to use in the subsequent official run.
+# For certain machine types the binary search result don't work well and will
+# cause kv errors and end the test, so we need to handle them differently.
+tpcc_files=""
+has_good_result=0
+if [[ special_type -eq 0 ]];
 then
-  f_inc=$f_warehouses
+  l=$f_active
+  r=$f_warehouses
+  while [[ $l -le $r ]]
+  do
+      mid=$((l + ((r - l)/2)))
+      report="${logdir}/tpcc-results-$mid.txt"
+      ./cockroach workload run tpcc \
+          --warehouses="$mid"  \
+          --active-warehouses="$mid" \
+          --ramp=1m --duration="1m" \
+          "${pgurls[@]}" > "$report"
+
+      # awk uses lexicographic order by default when doing comparison, we
+      # need to force it to use number comparison.
+      if [[ $(tail -1 "$report" | awk '{if($3+0 > 87){print "pass"}}') != "pass" ]];
```

Comment:
I think it's better to use the explicit cast `int($3)` which all recent awk variants support; e.g., same change was merged here: https://github.com/cockroachlabs/cloud-report/pull/91/files#diff-7d023a296a3338f43900d97c0a815b077a9919d020ca289267696cbc353af6eaR146

Reply:
fixed.